### PR TITLE
fix #1481.

### LINKED ISSUE #1481: 【neomake#Sh】Directory name invalid
### Expected behavior

Run  command asynchronously  in win32.

### Steps to reproduce

Minimal vimrc

```vim
if &compatible
  set nocompatible
endif

let &rtp=&rtp.','."c:/Users/tracyone/vimfiles/bundle/neomake"


colorscheme desert
filetype plugin indent on
syntax on
```

**Step**

```vim
:call neomake#Sh("dir ..")
:cw
```

#### Neomake debug information
Async support: 1
Current filetype: 
Windows: 1
[shell, shellcmdflag, shellslash]: ['C:\WINDOWS\system32\cmd.exe', '/c', 0]

##### Enabled makers
For the current filetype ("", used with :Neomake): None.
NOTE: the current buffer does not have a filetype.

For the project (used with :Neomake!): None.
NOTE: you can define g:neomake_enabled_makers to configure it.

Default maker settings:
 - remove_invalid_entries: 0
 - buffer_output: 1

##### Settings
```
g:neomake_place_signs = 1
```

#### :version
```

VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Jul 28 2017 22:02:02)
MS-Windows 64-bit GUI version 带 OLE 支持
包含补丁: 1-799
编译者 appveyor@APPVYR-WIN
巨型版本 带图形界面。  可使用(+)与不可使用(-)的功能:
+acl                +ex_extra           +multi_byte_ime/dyn -tag_any_white
+arabic             +extra_search       +multi_lang         +tcl/dyn
+autocmd            +farsi              +mzscheme/dyn       -termguicolors
+balloon_eval       +file_in_path       +netbeans_intg      -terminal
+browse             +find_in_path       +num64              -tgetent
++builtin_terms     +float              +ole                -termresponse
+byte_offset        +folding            +packages           +textobjects
+channel            -footer             +path_extra         +timers
+cindent            +gettext/dyn        +perl/dyn           +title
+clientserver       -hangul_input       +persistent_undo    +toolbar
+clipboard          +iconv/dyn          -postscript         +user_commands
+cmdline_compl      +insert_expand      +printer            +vertsplit
+cmdline_hist       +job                +profile            +virtualedit
+cmdline_info       +jumplist           +python/dyn         +visual
+comments           +keymap             +python3/dyn        +visualextra
+conceal            +lambda             +quickfix           +viminfo
+cryptv             +langmap            +reltime            +vreplace
+cscope             +libcall            +rightleft          +wildignore
+cursorbind         +linebreak          +ruby/dyn           +wildmenu
+cursorshape        +lispindent         +scrollbind         +windows
+dialog_con_gui     +listcmds           +signs              +writebackup
+diff               +localmap           +smartindent        -xfontset
+digraphs           +lua/dyn            +startuptime        -xim
+directx            +menu               +statusline         +xpm_w32
-dnd                +mksession          -sun_workshop       -xterm_save
-ebcdic             +modify_fname       +syntax             
+emacs_tags         +mouse              +tag_binary         
+eval               +mouseshape         +tag_old_static     
     系统 vimrc 文件: "$VIM\vimrc"
     用户 vimrc 文件: "$HOME\_vimrc"
 第二用户 vimrc 文件: "$HOME\vimfiles\vimrc"
 第三用户 vimrc 文件: "$VIM\_vimrc"
      用户 exrc 文件: "$HOME\_exrc"
  第二用户 exrc 文件: "$VIM\_exrc"
    系统 gvimrc 文件: "$VIM\gvimrc"
    用户 gvimrc 文件: "$HOME\_gvimrc"
第二用户 gvimrc 文件: "$HOME\vimfiles\gvimrc"
第三用户 gvimrc 文件: "$VIM\_gvimrc"
       defaults file: "$VIMRUNTIME\defaults.vim"
        系统菜单文件: "$VIMRUNTIME\menu.vim"
编译方式: cl -c /W3 /nologo  -I. -Iproto -DHAVE_PATHDEF -DWIN32  -DFEAT_CSCOPE  -DFEAT_NETBEANS_INTG -DFEAT_JOB_CHANNEL   -DFEAT_XPM_W32   -DWINVER=0x500 -D_WIN32_WINNT=0x500 /MP -DHAVE_STDINT_H /Ox /GL -DNDEBUG  /Zl /MT -DFEAT_OLE -DFEAT_MBYTE_IME -DDYNAMIC_IME -DFEAT_MBYTE -DFEAT_GUI_W32 -DFEAT_DIRECTX -DDYNAMIC_DIRECTX -DDYNAMIC_ICONV -DDYNAMIC_GETTEXT -DFEAT_TCL -DDYNAMIC_TCL -DDYNAMIC_TCL_DLL=\"tcl86.dll\" -DDYNAMIC_TCL_VER=\"8.6\" -DFEAT_LUA -DDYNAMIC_LUA -DDYNAMIC_LUA_DLL=\"lua53.dll\" -DFEAT_PYTHON -DDYNAMIC_PYTHON -DDYNAMIC_PYTHON_DLL=\"python27.dll\" -DFEAT_PYTHON3 -DDYNAMIC_PYTHON3 -DDYNAMIC_PYTHON3_DLL=\"python35.dll\" -DFEAT_MZSCHEME -I "C:\Program Files\Racket\include" -DMZ_PRECISE_GC -DDYNAMIC_MZSCHEME -DDYNAMIC_MZSCH_DLL=\"libracket3m_a0solc.dll\" -DDYNAMIC_MZGC_DLL=\"libracket3m_a0solc.dll\" -DFEAT_PERL -DPERL_IMPLICIT_CONTEXT -DPERL_IMPLICIT_SYS -DDYNAMIC_PERL -DDYNAMIC_PERL_DLL=\"perl524.dll\" -DFEAT_RUBY -DDYNAMIC_RUBY -DDYNAMIC_RUBY_VER=22 -DDYNAMIC_RUBY_DLL=\"x64-msvcrt-ruby220.dll\" -DFEAT_HUGE /Fd.\ObjGXOULYHTRZAMD64/ /Zi
链接方式: link /RELEASE /nologo /subsystem:windows /LTCG:STATUS oldnames.lib kernel32.lib advapi32.lib shell32.lib gdi32.lib  comdlg32.lib ole32.lib uuid.lib /machine:AMD64 gdi32.lib version.lib   winspool.lib comctl32.lib advapi32.lib shell32.lib  /machine:AMD64 /nodefaultlib libcmt.lib oleaut32.lib user32.lib  /nodefaultlib:lua53.lib  /STACK:8388608  /nodefaultlib:python27.lib /nodefaultlib:python35.lib   "C:\Tcl\lib\tclstub86.lib" WSock32.lib xpm\x64\lib\libXpm.lib /PDB:gvim.pdb -debug
```

#### :messages
```
简体中文消息维护者: Yuheng Xie <elephant@linux.net.cn>
```

call neomake#Sh whitout command's argument is work well.

### PR DESCRIPTION
**Pull request by tracyone**
remove unnecessary shellescape

### CODE REVIEW

File: autoload/neomake/compat.vim [hunk 1]
```diff
@@ -174,7 +174,7 @@ elseif neomake#has_async_support()  " Vim-async.
             if argv[0:len(prefix)-1] == prefix
                 return argv
             endif
-            return prefix.neomake#utils#shellescape(argv)
+            return argv
```
Comment (blueyed) [inline autoload/neomake/compat.vim]: `prefix` should be kept here, no?
It also break `tests/compat.vader` (you can try running it with `make tests/compat.vader`).

Everything after e.g. `sh -c` (or `cmd.exe /c`) should be shellescaped I think.

**Comment by tracyone** [comment]
@blueyed 

> prefix should be kept here, no?

no.
**argv** is  C:\WINDOWS\system32\cmd.exe" /c "dir .."

**prefix.argv** is C:\WINDOWS\system32\cmd.exe /c "C:\WINDOWS\system32\cmd.exe" /c "dir .."

**Comment by blueyed** [comment]
> argv is C:\WINDOWS\system32\cmd.exe" /c "dir .."

Really? (regarding the unbalanced quotes)

Anyway, the check before the return is meant to not use the shell twice, but seems to not work for you.

Keep in mind that you're using `NeomakeSh`, which is a special case since it uses the shell already, but the shell needs to be used in "normal" use cases, i.e. when running eslint etc, where typically `eslint.cmd` needs to be called (through a wrapping shell).

**Comment by blueyed** [comment]
Added another test in https://github.com/neomake/neomake/pull/1497.

Without the prefix check (first return) it would run it as `Starting async job: 'cmd.exe /c "cmd.exe /c ""dir .."""'.`.

**Comment by tracyone** [comment]
@blueyed why argv[0:len(prefix)-1] is not equal to prefix?

"C:\WINDOWS\system32\cmd.exe" /
C:\WINDOWS\system32\cmd.exe /c

**Comment by blueyed** [comment]
I see now.
Can you try https://github.com/neomake/neomake/pull/1499, please?

**Pull request closed without merging**